### DISCUSSION TITLE
Why does lsp-mode start so many workspaces?

### DISCUSSION THREAD

**Discussion by thomasheartman** [emacs-lsp/lsp-mode#3919] [Q&A] [answered]
It seems that when I start lsp-mode in a project, lsp-mode will also initialize a bunch of other projects that I haven't touched (and that I'm not going to touch). The other projects seem entirely unrelated and I do not have any open buffers in them. 

Why is this happening and can I disable it? To be fair, I don't know whether it's causing any performance issues, but it does start a bunch of language servers, so I  imagine it's not great anyway 🤷🏼 

Also, is it possible to stop a language server workspace from the lsp session overview? lsp-describe-session shows all the workspaces (including the aforementioned ones that I don't  think should be active), and I'd like to be able to shut them down from there.

I'd be happy to provide more context, but this is about all the information I have at the moment. If there's any steps I can take to provide more data, let me know 😄 

Any input would be much appreciated! Thank you 🙏🏼

**Comment by yyoncho** [comment, marked as answer]
When the server is multiroot(supports several folders) we open all of the folders with the same server and we persist that. Some servers have a flag to disable that behavior, for others it is not exposed.

**Reply by thomasheartman** [reply]
Oh, interesting! How can  I tell whether a server is multiroot? And is there a unified way to disable the behavior if the server exposes it?

**Reply by yyoncho** [reply]
Check https://github.com/emacs-lsp/lsp-pyright/blob/master/lsp-pyright.el#L138

**Reply by yyoncho** [reply]
And this: https://emacs-lsp.github.io/lsp-mode/page/faq/#how-do-i-force-lsp-mode-to-forget-the-workspace-folders-for-multi-root-servers-so-the-workspace-folders-are-added-on-demand

**Reply by thomasheartman** [reply]
Ooh, that multi-root disable snippet seems to have done the trick 😄  Thank you! 🙏🏼 

What about my second question: is there a way to quit workspaces from the lsp session overview? It'll be less useful if all the workspaces don't start automatically, but I'm still curious.

**Reply by yyoncho** [reply]
I guess you are looking for `lsp-workspace-folders-remove`?

**Reply by thomasheartman** [reply]
Well, maybe. 

When I invoke lsp-describe-session, I'm taken to the session overview. The overview lists many of (all?) the workspaces that I have started. Some are active and running, others are not.

I would like to be able to quit a workspace lsp session from that screen. For instance, if `/project/a` has a running session, I would like to move the cursor to that line and press `k` (or whatever) and have that lsp server quit. I don't necessarily want to remove the project from the overview.

Does that make sense? I'm also not sure my terminology (session vs workspace vs project) is spot on here, so if there's any ambiguity, that might be the reason why.

**Reply by yyoncho** [reply]
You can quit the server using lsp-workspace-shutdown.

**Reply by thomasheartman** [reply]
Yes, but from what I can tell, that requires you to be in a buffer that has that server active. If you don't have any buffers in `/project/a`, you'd have to open one and then invoke `lsp-workspace-shutdown` from there. If I try and invoke it in the lsp session buffer, I get a message saying "No workspaces associated with the current buffer". 

Is there no way to quit servers from the lsp session view?

**Reply by yyoncho** [reply]
> Is there no way to quit servers from the lsp session view?

No.

**Reply by thomasheartman** [reply]
Got it. Thanks for all your help ☺️
